File: src/Day06.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.Map.Entry;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static javautils.JavaUtils.*;

public class Day06 {

  public static void main(String[] args) throws Exception {
    if (args.length == 0) {
      List<String> testInput = Files.readAllLines(Paths.get("src/06t.txt"));
      assert part1(testInput) == 41 : "Expected 41 but got " + part1(testInput);

      List<String> input = Files.readAllLines(Paths.get("src/06.txt"));
      System.out.println(part1(input));
      assert part2(testInput) == 6 : "Expected 6 but got " + part2(testInput);
      System.out.println(part2(input));
    } else {
      List<String> input = Files.readAllLines(Paths.get(args[0]));
      System.out.println(part1(input));
      System.out.println(part2(input));
    }
  }

  private static int part1(List<String> input) {
    JGrid<Character> grid = toGrid(input);
    JPoint2D startingPoint = getStartingPoint(grid);
    return walk(grid, startingPoint).size();
  }

  private static int part2(List<String> input) {
    JGrid<Character> grid = toGrid(input);
    JPoint2D startingPoint = getStartingPoint(grid);
    Set<JPoint2D> initialWalk = new HashSet<>(walk(grid, startingPoint));

    return (int) initialWalk
        .stream()
        .filter(toReplace -> {
          var modifiedGrid = copyAndReplace(grid, toReplace, '#');
          JPoint2D curPoint = startingPoint;
          JPoint2D curDirection = UP;
          Set<Entry<JPoint2D, JPoint2D>> visited = new HashSet<>();
          while (true) {
            var next = next(modifiedGrid, curPoint, curDirection);
            if (next == null) return false;
            if (!visited.add(next)) break;
            curPoint = next.getKey();
            curDirection = next.getValue();
          }
          return true;
        })
        .count();
  }

  private static Entry<JPoint2D, JPoint2D> next(JGrid<Character> grid, JPoint2D point, JPoint2D direction) {
    JPoint2D nextDirection = direction;
    if (Objects.equals(grid.get(point.plus(direction)), '#')) {
      nextDirection = turn(direction);
      if (Objects.equals(grid.get(point.plus(nextDirection)), '#')) {
        nextDirection = turn(nextDirection);
      }
    }
    JPoint2D nextPoint = point.plus(nextDirection);
    return grid.contains(nextPoint) ? Map.entry(nextPoint, nextDirection) : null;
  }

  private static JPoint2D turn(JPoint2D direction) {
    if (UP.equals(direction)) return RIGHT;
    else if (RIGHT.equals(direction)) return DOWN;
    else if (DOWN.equals(direction)) return LEFT;
    else return UP; // LEFT case
  }

  private static Set<JPoint2D> walk(JGrid<Character> grid, JPoint2D startingPoint) {
    Set<JPoint2D> visited = new HashSet<>();
    visited.add(startingPoint);
    JPoint2D nextPoint = startingPoint;
    JPoint2D direction = UP;

    while (true) {
      Entry<JPoint2D, JPoint2D> next = next(grid, nextPoint, direction);
      if (next == null) break;
      nextPoint = next.getKey();
      direction = next.getValue();
      visited.add(nextPoint);
    }

    return visited;
  }

  private static JPoint2D getStartingPoint(JGrid<Character> grid) {
    return StreamSupport.stream(grid.spliterator(), false)
        .filter(e -> grid.get(e) == '^')
        .map(e -> new JPoint2D(e.x(), e.y()))
        .findFirst()
        .orElseThrow();
  }

  private static JGrid<Character> toGrid(List<String> input) {
    List<List<Character>> gridData = input
        .stream()
        .map(line -> line.chars().mapToObj(c -> (char) c).collect(Collectors.toList()))
        .collect(Collectors.toList());
    return new JGrid<>(gridData);
  }

  private static JGrid<Character> copyAndReplace(JGrid<Character> grid, JPoint2D toReplace, char ch) {
    var newGrid = new JGrid<>(new HashMap<>(grid.data()));
    newGrid.set(toReplace, ch);
    return newGrid;
  }
}

File: src/Day06.kt
import kotlin.io.path.Path
import kotlin.io.path.readText


fun main(args: Array<String>) {
    fun Grid<Char>.next(point: Point2D, direction: Point2D): Pair<Point2D, Point2D>? {
        fun turn(direction: Point2D) = when (direction) {
            UP -> RIGHT
            RIGHT -> DOWN
            DOWN -> LEFT
            LEFT -> UP
            else -> error("Unsupported direction: $direction")
        }

        var nextDirection = direction
        if (this[point + direction] == '#') {
            nextDirection = turn(direction)
            // turn two times!!
            if (this[point + nextDirection] == '#') nextDirection = turn(nextDirection)
        }
        val nextPoint = point + nextDirection
        return if (nextPoint in this) return (nextPoint to nextDirection) else null
    }

    fun Grid<Char>.walk(startingPoint: Point2D): HashSet<Point2D> {
        val visited = hashSetOf(startingPoint)
        var nextPoint: Point2D = startingPoint
        var direction = UP
        while (true) {
            val (first, second) = next(nextPoint, direction) ?: break
            nextPoint = first
            direction = second
            visited.add(nextPoint)
        }
        return visited
    }

    fun part1(input: List<String>): Int {
        val grid = input.toGrid()
        val startingPoint = grid.data.entries.first { it.value == '^' }.key
        return grid.walk(startingPoint).size
    }

    fun part2(input: List<String>): Int {
        val grid = input.toGrid()
        val startingPoint = grid.data.entries.first { it.value == '^' }.key
        val initialWalk = grid.walk(startingPoint).toSet()
        return initialWalk
            .filter { toReplace ->
                val grid2 = grid.copy(data = grid.data.toMutableMap().also { it[toReplace] = '#' })
                var curPoint = startingPoint
                var curDirection = UP
                val visited = hashSetOf(curPoint to curDirection)
                while (true) {
                    val tmp1 = grid2.next(curPoint, curDirection) ?: return@filter false
                    if (!visited.add(tmp1)) break
                    curPoint = tmp1.first
                    curDirection = tmp1.second
                }
                true
            }
            .count()
    }


    // Or read a large test input from the `src/Day01_test.txt` file:
    if (args.isEmpty()) {
        val testInput = readInput("06t")
        check(part1(testInput) == 41) { "Expected 41 but got ${part1(testInput)}" }

        // Read the input from the `src/Day01.txt` file.
        val input = readInput("06")
        part1(input).println()
        check(part2(testInput) == 6) { "Expected 6 but got ${part2(testInput)}" }
        part2(input).println()
    } else {
        val input = Path(args[0]).readText().trim().lines()
        part1(input).println()
        part2(input).println()
    }
}


